### PR TITLE
Fix KnownSafety optimization bugs in ARCSequenceOpts

### PR DESCRIPTION
**Pull request by meg-gupta**
ARCSequenceOpts does KnownSafety optimization where 'KnownSafe' retain/release pairs are deleted.
A retain is 'KnownSafe' if during top down dataflow, it is preceded by another retain or non-arc use on the same RCIdentity.
A release is 'KnownSafe' if during bottom up dataflow, it is preceded by another release or non-arc use on the same RCIdentity.
This PR fixes the following two bugs in KnownSafety optimization.

a) KnownSafety bug in ARCSequenceOpts without loop support

    1: retain_value S
    2: retain_value S

    3: strong_release S.subobject
    4: strong_retain S.subobject
    5: release_value S
    6: release_value S

2 and 5 are marked as 'KnownSafe'. And currently they get paired and optimized away. 2 and 5 pairing is incorrect, this can be seen clearly expanding retain_value and release_value.
   
    1e: strong_retain S.subobject
    2e: strong_retain S.subobject
    3: strong_release S.subobject
    4: strong_retain S.subobject
    5e: strong_release S.subobject
    6e: strong_release S.subobject

The correct pairings are 2e and 3, 4 and 5e. Since RCIdentity distinguishes between S and S.subobject we can end
up with incorrect pairings and optimization with KnownSafety. This is true for any aliasing RCIdentities.

b) KnownSafety bug in ARCSequenceOpts with loop support

    bb0:
       1: strong_retain S
       2: strong_retain S
    loop:
        3: strong_release S
        4: strong_retain S
        5: cond_br undef, loop, bb2
    bb2:
       6: strong_release S
       7: strong_release S

2 and 6 are marked as 'KnownSafe'. And currently get paired and optimized away. This is incorrect and the correct pairings are 2 and 3 and 4 and 6. ARCLoopOpts computes loop summaries and uses it to only compute refcount states for CodeMotionSafe optimization in ARCSequenceOpts. The instructions in the loop have no effect on the already computed 'KnownSafe' states outside the loop.

With this PR, while doing bottom up dataflow, if we encounter an unmatched retain instruction, that can pair with a 'KnownSafe' already visited release instruction, we turn off KnownSafety if the two
RCIdentities mayAlias.
Similarly, during top down dataflow, if we encounter an unmatched release instruction, that can pair with a 'KnownSafe' already visited retain instruction, we turn off KnownSafety if the two RCIdentities mayAlias. This fixes the bugs in both ARCSequenceOpts without loop support and with loop support.

Fixes rdar://67720198 and rdar://27775427

### CODE REVIEW

**Review by gottesmm — COMMENTED** [review]
There is a bunch of refactoring going on here. Can you split that out in a separate PR and land that separately?

File: lib/SILOptimizer/ARC/ARCRegionState.cpp [hunk 1]
```diff
@@ -156,13 +156,21 @@ void ARCRegionState::mergePredTopDown(ARCRegionState &PredRegionState) {
...
 
-  auto II = State.summarizedinterestinginsts_rbegin();
-  auto IE = State.summarizedinterestinginsts_rend();
+  SILBasicBlock &BB = *R->getBlock();
+  BottomUpDataflowRCStateVisitor<ARCRegionState> DataflowVisitor(
+      RCIA, EAFI, *this, FreezeOwnedArgEpilogueReleases, IncToDecStateMap,
+      SetFactory);
+
+  auto II = this->summarizedinterestinginsts_rbegin();
+  auto IE = this->summarizedinterestinginsts_rend();
```
Comment (gottesmm) [inline lib/SILOptimizer/ARC/ARCRegionState.cpp:173]: Please remove the usage of this.

**Comment by gottesmm** [comment]
Just reading the description, this is incorrect. We only require bottom up that there is a post-dominating use that requires liveness that will not be removed.

**Comment by meg-gupta** [comment]
@gottesmm I edited the description of 'KnownSafety' to also include non-arc use.

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci benchmark

**Review by atrick — COMMENTED** [review]
I'm still trying to prove to myself it's correct and efficient.

I was initially worried that this would fail to optimize a more "normal" case like this:
```
1: retain_value S
2: retain_value S
3: strong_retain S.subobject
4: strong_release S.subobject
5: release_value S
6: release_value S
```
Maybe that case will actually sill be optimized with your change? Anyway, I won't argue strongly that this pass needs to handle all cases since SemanticARCOpts can handle it eventually.

Will we correctly handle a case like this?
```
1: retain_value S
2: retain_value S
3: strong_release S.subobject
4: release_value S
5: strong_retain S.subobject
6: release_value S
```

**Comment by meg-gupta** [comment]
@atrick We will still optimize the "normal" case. Test `$matched_rr_subobject` in `arcsequenceopts_knownsafetybugs.sil` has this pattern.

**Comment by meg-gupta** [comment]
In:

    1: retain_value S
    2: retain_value S
    3: strong_release S.subobject
    4: release_value S
    5: strong_retain S.subobject
    6: release_value S

2 and 4 get paired and optimized away due to CodeMotionSafe. This fix turns off the KnownSafety on 2 when it sees 3.

**Comment by atrick** [comment]
I'm trying to understand why this known safety fix is looking for retains. Let's say code-motion safe doesn't apply. So what about  this?:
```
1: retain_value S
2: retain_value S
3: strong_release S.subobject
4: use(S)
5: release_value S
6: strong_retain S.subobject
7: release_value S
```

**Comment by meg-gupta** [comment]
@atrick Similar to previous example, 3 turns off KnownSafety on 2. And nothing gets optimized if we consider only 'KnownSafety'. 
In bottom up traversal, the fix only looks at unmatched retains that could potentially match with already visited KnownSafe RCIdentity.

**Comment by atrick** [comment]
For efficiency, RefCountInstToMatchedMap can just be a set of unmatched instructions.

**Comment by atrick** [comment]
When we summarize the loop in the RefCountInstToMatchedMap. It assumes that the IncToDecStateMap finds any unmatched pair on any paths through the loop.

LoopARCSequenceDataflowEvaluator::mergeSuccessors clears the data flow state at any non-local successor, so it should be conservative.

This only time this doesn't happen is with "allowsleaks" exist, which reach an unreachable before any side-effect instructions.

To be safe, add tests cases for when the release is on the loop tail; add a test:
```
loop:
  retain
  cond_br exit
  release
  br loop

loop:
  retain
  cond_br exit-to-unreachable
  release
  br loop
```

**Comment by atrick** [comment]
When computing the loop summary, the code assumes that dataflow can be applied to each "interesting" instruction in order to produce a summary. In general this is not sound because loops iterate and the order is meaningless. This means that we're relying on `updateForDifferentLoopInst` to apply data flow so conservatively that the order doesn't matter. It's not clear that's actually happening. We should clearly comment that requirement and verify the assumption under some verification flag:

In ARCRegionState::processLoopBottomUp
```
  // For each state that we are currently tracking, apply our summarized
  // instructions to it.
  for (auto &OtherState : getBottomupStates()) {
    if (!OtherState.hasValue())
      continue;

    for (auto *I : State->getSummarizedInterestingInsts()) {
      OtherState->second.updateForDifferentLoopInst(I, AA);
      OtherState->second.checkAndResetKnownSafety(
          I, OtherState->first, checkIfRefCountInstIsMatched, RCIA, AA);
    }
#ifndef NDEBUG
    if (verifyLoopARCSummary) {
      auto newOtherState = OtherState
      for (auto *I : State->getSummarizedInterestingInsts()) {
        OtherState->second.updateForDifferentLoopInst(I, AA);
        OtherState->second.checkAndResetKnownSafety(
            I, OtherState->first, checkIfRefCountInstIsMatched, RCIA, AA);
      }
      assert(newOtherState == OtherState);
    }
#endif
  }
```

**Review by atrick — APPROVED** [review]
I asked for some test cases and comments. Otherwise this looks good to me. Thanks!

Regarding the loop data-flow verification, that's a separate issue from this PR so can be done in a follow-up.

**Comment by meg-gupta** [comment]
@swift-ci Please Test Source Compatibility

**Comment by meg-gupta** [comment]
@swift-ci Please Test Source Compatibility

**Comment by meg-gupta** [comment]
Addressed review comments.

**Comment by meg-gupta** [comment]
sck debug failures are pre-existing

**Pull request merged**